Mathlib/Analysis/Analytic/OfScalars.lean

```diff
@@ -0,0 +1,120 @@
+convergence is equal to $\lim_{n \to \infty} \|c_{n+1}\|\cdot\|c_n\|^{-1}$ as long as it is not `0`.
+-/
+
+namespace FormalMultilinearSeries
+
+section Field
+
+variable {𝕜 : Type*} (E : Type*) [Field 𝕜] [Ring E] [Algebra 𝕜 E] [TopologicalSpace E]
+  [TopologicalRing E] (c : ℕ → 𝕜)
+
+/-- Formal power series of `∑ cᵢ • xⁱ` for some scalar field `𝕜` and ring algebra `E`-/
+def ofScalars : FormalMultilinearSeries 𝕜 E E :=
+  fun n ↦ c n • ContinuousMultilinearMap.mkPiAlgebraFin 𝕜 n E
+
+variable {E}
+
+/-- The sum of the formal power series. Takes the value `0` outside the radius of convergence. -/
+noncomputable def ofScalarsSum (x : E) := (ofScalars E c).sum x
+
+theorem ofScalars_apply_eq (x : E) (n : ℕ) :
+    ofScalars E c n (fun _ ↦ x) = c n • x ^ n := by
+  simp [ofScalars]
+
+/-- This naming follows the convention of `NormedSpace.expSeries_apply_eq'`. -/
+theorem ofScalars_apply_eq' (x : E) :
+    (fun n ↦ ofScalars E c n (fun _ ↦ x)) = fun n ↦ c n • x ^ n := by
+  simp [ofScalars]
+
+theorem ofScalars_sum_eq (x : E) : ofScalarsSum c x =
+    ∑' n, c n • x ^ n := tsum_congr fun n => ofScalars_apply_eq c x n
+
+theorem ofScalarsSum_eq_tsum : ofScalarsSum c =
+    fun (x : E) => ∑' n : ℕ, c n • x ^ n := funext (ofScalars_sum_eq c)
+
+@[simp]
+theorem ofScalars_op [T2Space E] (x : E) :
+    ofScalarsSum c (MulOpposite.op x) = MulOpposite.op (ofScalarsSum c x) := by
+  simp [ofScalars, ofScalars_sum_eq, ← MulOpposite.op_pow, ← MulOpposite.op_smul, tsum_op]
+
+@[simp]
+theorem ofScalars_unop [T2Space E] (x : Eᵐᵒᵖ) :
+    ofScalarsSum c (MulOpposite.unop x) = MulOpposite.unop (ofScalarsSum c x) := by
+  simp [ofScalars, ofScalars_sum_eq, ← MulOpposite.unop_pow, ← MulOpposite.unop_smul, tsum_unop]
+
+@[simp]
+theorem ofScalars_eq_zero [Nontrivial E] (n : ℕ) : ofScalars E c n = 0 ↔ c n = 0 := by
+  rw [ofScalars, smul_eq_zero (c := c n) (x := ContinuousMultilinearMap.mkPiAlgebraFin 𝕜 n E)]
+  refine or_iff_left (ContinuousMultilinearMap.ext_iff.1.mt <| not_forall_of_exists_not ?_)
+  use fun _ ↦ 1
+  simp
+
+end Field
+
+section Normed
+
+open Filter
+open scoped Topology
+
+variable {𝕜 : Type*} (E : Type*) [NontriviallyNormedField 𝕜] [NormedRing E]
+    [NormedAlgebra 𝕜 E] (c : ℕ → 𝕜) (n : ℕ)
+
+theorem ofScalars_norm_eq_mul :
+    ‖ofScalars E c n‖ = ‖c n‖ * ‖ContinuousMultilinearMap.mkPiAlgebraFin 𝕜 n E‖ := by
+  rw [ofScalars, norm_smul (c n) (ContinuousMultilinearMap.mkPiAlgebraFin 𝕜 n E)]
+
+theorem ofScalars_norm_le (hn : n > 0) : ‖ofScalars E c n‖ ≤ ‖c n‖ := by
+  simp only [ofScalars_norm_eq_mul]
+  exact (mul_le_of_le_one_right (norm_nonneg _)
+    (ContinuousMultilinearMap.norm_mkPiAlgebraFin_le_of_pos hn))
+
+@[simp]
+theorem ofScalars_norm [NormOneClass E] : ‖ofScalars E c n‖ = ‖c n‖ := by
+  simp [ofScalars_norm_eq_mul]
+
+/-- The radius of convergence of a scalar series is the inverse of the ratio of the norms of the
+coefficients. -/
+theorem ofScalars_radius_of_tendsto [NormOneClass E] {r : NNReal} (hr : r ≠ 0)
```

Comment:
How about we call this:
```suggestion
theorem ofScalars_radius_eq_inv_of_tendsto [NormOneClass E] {r : NNReal} (hr : r ≠ 0)
```
and then add the following convenience lemma below:
```lean
theorem ofScalars_radius_eq_of_tendsto [NormOneClass E] {r : NNReal} (hr : r ≠ 0)
    (hc : Tendsto (fun n ↦ ‖c n‖ / ‖c n.succ‖) atTop (𝓝 r)) :
      (ofScalars E c).radius = ENNReal.ofNNReal r := by
  suffices Tendsto (fun n ↦ ‖c n.succ‖ / ‖c n‖) atTop (𝓝 r⁻¹) by
    convert ofScalars_radius_eq_inv_of_tendsto E c (inv_ne_zero hr) this
    simp
  convert (continuousAt_inv₀ <| NNReal.coe_ne_zero.mpr hr).tendsto.comp hc
  simp
```
(which should actually be better suited to your application)